src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.AnalyzerStateData.cs

```diff
@@ -56,6 +57,7 @@ public void ResetToReadyState()
 
             public virtual void Free()
             {
+                Debug.Assert(StateKind != StateKind.InProcess);
```

Comment:
>StateKind != StateKind.InProcess [](http://example.com/codeflow?start=29&length=32)

Does it make sense to make this assert stronger and check that we are in ```FullyProcessed``` state?

Reply:
Actually the `StateKind` would be `ReadyToBeProcessed` here, I'll change the assert here accordingly. `StateKind.FullyProcessed` is only used for singletons : https://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalysisState.StateKind.cs,2777a03ff8c49043,references to reduce allocations. When transitioning analysis to complete state, we use the below code pattern: https://github.com/dotnet/roslyn/blob/440b6fc370f0ddcea68b445e183f1383c0d1bc6e/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.PerAnalyzerState.cs#L281-L286